### PR TITLE
5353 iterate on revincludes and includes does not return correct resources when used with non iterate revincludes

### LINKED ISSUE #5353: Iterate on revincludes and includes does not return correct resources when used with non iterate revincludes
**Describe the bug**
When using revincludes and includes with iterate, while also using revincludes without iterate, the result omits some resources that should be included.

**To Reproduce**
1. Create an EpisodeOfCare & a Patient
2. Create an Encounter that reference both the EpisodeOfCare and the Patient
3. Create a Task that reference the Encounter
4. Run the following query: `EpisodeOfCare?identifier=system1|value1&_revinclude=Encounter:episode-of-care&_include:iterate=Encounter:patient&_revinclude:iterate=Task:encounter`
5. See only 3 resources returned in the results

**Expected behavior**
Should return all 4 resources

**Environment (please complete the following information):**
 - HAPI FHIR Version: 6.9.9
 - OS: MacOS
 - Browser: chrome

### PR DESCRIPTION
**Pull request by StevenXLi**
- replace the check for revinlcudes with :iterate (or :recurse for earlier versions of FHIR) with separating of the search by revinclude, include, revinclude with iterate, and include with iterate.
- File to review: Search.java, PersistedJpaBundleProvider.java, and ResourceProviderRevIncludeTest.java
- Non-impact change: added generated docs

closes #5353

### CODE REVIEW

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/MdmMetrics.java [hunk 1]
```diff
@@ -1,3 +1,22 @@
+/*-
+ * #%L
+ * HAPI FHIR - Master Data Management
+ * %%
+ * Copyright (C) 2014 - 2023 Smile CDR, Inc.
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/MdmMetrics.java:5]: I've seen james do the same. Maybe we want some sort of process for this. Seems like this type of change should be done by the person who wrote the class.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java [hunk 2]
```diff
@@ -463,72 +462,70 @@ protected List<IBaseResource> toResourceList(
 		if (mySearchEntity.getSearchType() == SearchTypeEnum.SEARCH) {
 			Integer maxIncludes = myStorageSettings.getMaximumIncludesToLoadPerPage();
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java:463]: I you are talking about the SearchBuilder call, the null check is handled further down the processing stack. It's redundant to do it here. line 1399 SearchBuilder.java

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5353-iterate-on-revincludes-and-includes-does-not-return-correct-resources-when-used-with-non-iterate-revincludes.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 5353
+jira: SMILE-7451
+title: "Previously, using revincludes without iterate and revincludes and includes with iterate together,
+will omit some resources that should be included. This issue has been fixed."
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5353-iterate-on-revincludes-and-includes-does-not-return-correct-resources-when-used-with-non-iterate-revincludes.yaml]: Next time of these types of change requests you can just use the add suggestion button. 
image

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/MdmMetrics.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/MdmMetrics.java:5]: 💯

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5353-iterate-on-revincludes-and-includes-does-not-return-correct-resources-when-used-with-non-iterate-revincludes.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5353-iterate-on-revincludes-and-includes-does-not-return-correct-resources-when-used-with-non-iterate-revincludes.yaml]: Thanks.  Didn't know how to do this.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java:463]: 👍

**Pull request merged**